designs/preflight-checks.md

```diff
@@ -0,0 +1,163 @@
+* Ensure design is extensible to easily incorporate additional validations in the future
+
+**Not in scope**
+
+* Creating an exhaustive list of additional preflight checks 
+* Large scale changes to documentation
+
+**Future scope**
+
+* Additional general preflight validation checks
+* Additional in-depth validation for additional providers
+* Validation for the `upgrade cluster` command
+
+## Overview of solution
+
+EKS Anywhere includes extensive preflight checks already run as part of the create cluster command. These validations will be incorporated into an independent `validate` command. This command will initially be introduced as a subcommand under the `exp` command to indicate this is an experimental command. The `validate` command will be easily extensible for additional preflight-specific validations and different provider configurations.  
+
+The existing validations should continue to be run as part of  `create cluster`  to prevent any possible regressions or invalid cluster creation from being overlooked.
+
+### Architectural design
+
+#### Validations Overview
+
+The `validate` command architecture will differ from the `create cluster` workflow by first creating all the necessary dependencies to run validations, then running all validations the aggregated together in the validatevalidations package. The initial dependency generation will be a prerequisite to run validations and any failure during this will result in an err response for the command. 
+
+Running the validations together after dependency generation will allow better aggregation for the validation results, as well as ensure that the `validate` runs all validations possible instead of returning early on individual validation failure. 
+
+Ideally, the dependency generation process should be separate from the their validation to enable this architecture and allow the generation and validations to be called independently and or together as needed.
+
+#### Sequence Diagram
+
+![preflight checks sequence diagram](images/preflight-checks-sequence.png)
+#### Validation Location
+
+The `validate` command includes validations from the following locations originally based off the `create cluster` command validations. This does not include all code dependencies for the `validate` command. 
+
+```
+
+`.`
+`└──`` eks``-``anywhere`
+`    ``├──`` cmd``/``eks``-``a``/``cmd``
+    │ ├── experiment.go 
+    │ ├── validate.go 
+    │ └── validatecreatecluster.go
+``    ``└──`` pkg`
+`        ``├──`` addonmanager``/``addonclients`
+`        ``│``   ``└──`` fluxaddonclient``.``go`
+`        ``├──`` api``/``v1alpha1`
+`        ``│``   ``└──`` cluster``.``go`
+`        ``├──`` providers`
+`        ``│``   ``├──`` cloudstack`
+`        ``│``   ``├──`` docker`
+`        ``│``   ``├──`` snow`
+`        ``│``   ``├──`` tinkerbell`
+`        ``│``   ``└──`` vsphere`
+`        ``└──`` validations`
+`            ``├── validatevalidations`
```

Comment:
in this package new? The name is a bit weird

Reply:
can they just be in `validations`?
Or maybe depending on what they are validating they can go in there own packages? Like if it's a docker validation, in a docker package, etc.

@chrisdoherty4 is better than be at naming packages

Reply:
If its a set of validation helpers `validation` is fine. Initially a `validation.go` is fine and if we start finding groups of algorithms like 'network validation' then we can isolate that in its own source file down the road.

I wouldn't follow the `createvalidations` thing - that's rather odd.

Reply:
targeted validation helpers under `validation` folder but ended up making something pretty similar to the `createvalidations` flow - which as Chris mentioned may not be what we're shooting for. Will follow up to see what makes sense to move/restructure.

Reply:
Updated to move away from the `createvalidations` flow, currently this just consists of a `createcluster` pkg in validations that runs the associated validations in sequence.